File: app/src/main/java/com/dominikgaller/alpinebooking/booking/core/domain/tourbooking/AvailableCapacity.java
package com.dominikgaller.alpinebooking.booking.core.domain.tourbooking;

/**
 * Value object representing the number of open spots for a tour/date at booking time.
 *
 * <p>Stored as a snapshot in the aggregate. Invariant: value must be >= 0.
 *
 * <p>SDD: See {@code documentation/domain/aggregate-tour-booking.spec.md}, section 2a.
 */
public record AvailableCapacity(int value) {

    public AvailableCapacity {
        if (value < 0) {
            throw new IllegalArgumentException(
                    "AvailableCapacity must be >= 0, was: " + value);
        }
    }
}

File: src/main/kotlin/com/example/service/booking/core/domain/tourbooking/AvailableCapacity.kt
package com.example.service.booking.core.domain.tourbooking

/**
 * Open spots on a tour, as reported by the availability system at a point in time.
 *
 * A snapshot, not a live figure. It is stored on the aggregate so that a later decision
 * can be judged against the capacity that was actually observed when it was made.
 *
 * Uses [IllegalArgumentException] deliberately: this value never arrives from a command,
 * only from the `AvailabilityChecker` outport or from reconstitution, so a negative value
 * is a programmer error rather than bad input (`coding-style.definition.md` § 6.2).
 *
 * SDD: see `documentation/domain/aggregate-tour-booking.spec.md`.
 */
data class AvailableCapacity(
    val value: Int,
) {
    init {
        require(value >= 0) { "AvailableCapacity must be >= 0, was: $value" }
    }
}
